src/canopy/chat_engine/query_generator/condensed.py

```diff
@@ -0,0 +1,139 @@
+from canopy.chat_engine.prompt_builder import PromptBuilder
+from canopy.chat_engine.query_generator import QueryGenerator, LastMessageQueryGenerator
+from canopy.llm import BaseLLM, AnyscaleLLM
+from canopy.models.data_models import Messages, Query, UserMessage
+from canopy.models.api_models import ChatResponse
+
+logger = logging.getLogger(__name__)
+
+SYSTEM_PROMPT = """
+You are an expert on formulating a search query for a search engine, to assist
+ in responding to the user's question.
+
+Given the following conversation, create a standalone question summarizing
+ the user's last question, in its original language.
+
+Reply to me in JSON in this format:
+
+ {"question": {The question you generated here}}.
+
+   Example:
+
+        User: What is the weather today?
+
+        Expected Response:
+            ```json
+            {"question": "What is the weather today?"}
+            ```
+
+    Example 2:
+
+        User: How should I wash my white clothes in the laundry?
+        Assistant: Separate from the colorful ones, and use a bleach.
+        User: Which temperature?
+
+        Expected Response:
+         ```json
+         {"question": "What is the right temperature for washing white clothes?"}
+         ```
+
+
+   Do not try to answer the question; just try to formulate a question representing the user's question.
+   Do not return any other format other than the specified JSON format and keep it really short.
+
+"""  # noqa: E501
+
+USER_PROMPT = "Return only a JSON that has 'question' as a key and the value."
+
+
+class ExtractionException(ValueError):
+    pass
+
+
+class CondensedQueryGenerator(QueryGenerator):
+    _DEFAULT_COMPONENTS = {
+        "llm": AnyscaleLLM,
+    }
+
+    def __init__(self,
+                 *,
+                 llm: Optional[BaseLLM] = None):
+        """
+             Tries to build a question out of all the chat history.
+             In order to do that, the system first tries to make LLM return
+             a JSON object with a key "question" and the value representing
+             the question the user was intending to ask. If LLM response
+             cannot be parsed it falls back to last message query generator,
+             meaning it returns the last message of the history as a query.
+        """
+        self._llm = llm or self._DEFAULT_COMPONENTS["llm"]()
+        self._system_prompt = SYSTEM_PROMPT
+        self._prompt_builder = PromptBuilder(HistoryPruningMethod.RAISE, 2)
+        self._last_message_query_generator = LastMessageQueryGenerator()
+
+        # Define a regex pattern to find the JSON object with the key "question"
+        self._question_regex = re.compile(r'{\s*"question":\s*"([^"]+)"\s*}')
+
+    def generate(self,
+                 messages: Messages,
+                 max_prompt_tokens: int) -> List[Query]:
+
+        # Add a user message at the end; that helps us return a JSON object.
+        new_history = (
+                messages +
```

Comment:
We should probably make this configurable - e.g. by adding a `append_instruction_reminder_to_history: bool = True` parameter. (Feel free to find a shorter name... 😄)

It might be that some LLMs don't respond well to this additional message, and it only confuses them.

Reply:
OpenAI and Anyscale LLMs respond it quite well, I think in any case there is a lot of logic tied to this query generator so not sure customizing only that part will give us more flexibility. We will probably put a parameter that no one will use. In the future maybe we can improve this to template the user prompt & system prompt and give complete flexibility to the user but for now I wanna keep it as it is. LMK what you think

Reply:
Ok, agreed.